### PR TITLE
Issue #83 add suport of generic containers to VOI

### PR DESCRIPTION
**Pull request by dv14**
add support of generic index based containers to VOI instead of std::vector according to issue #83

### CODE REVIEW

File: modules/distance/k-random/VOI.cpp [hunk 1]
```diff
@@ -27,11 +28,12 @@ Copyright (c) 2019 Panda Team
...
 
 namespace {
-    template <typename T>
-    void add_noise(std::vector<std::vector<T>>& data)
+    template <typename C>
+    void add_noise(C & data)
     {
         std::random_device rd;
         std::mt19937 gen(rd());
+        using T = type_traits::underlaying_type_t<C>;
```
Comment (jura-gresko) [inline modules/distance/k-random/VOI.cpp]:
```suggestion
        using T = type_traits::underlaying_type_t<C>;
        std::random_device rd;
        std::mt19937 gen(rd());
```

File: modules/distance/k-random/VOI.cpp [hunk 2]
```diff
@@ -69,16 +71,20 @@ namespace {
         }
         std::cout << "]" << std::endl;
     }
-    template <typename T>
+    template <typename C1, typename C2, typename C3>
     void combine(
```
Comment (jura-gresko) [inline modules/distance/k-random/VOI.cpp]: can combine return std::vector<vector> here? as unique

File: modules/distance/k-random/VOI.hpp [hunk 3]
```diff
@@ -112,10 +123,18 @@ struct VOI {
...
      */
-    template <template <class, class> class Container, class Allocator_inner, class Allocator_outer, class El>
-    typename std::enable_if<!std::is_integral<El>::value, V>::type  // only real values are accepted
-    operator()(const Container<Container<El, Allocator_inner>, Allocator_outer>& a,
-        const Container<Container<El, Allocator_inner>, Allocator_outer>& b) const;
+    // template <template <class, class> class Container, class Allocator_inner, class Allocator_outer, class El>
+    // typename std::enable_if<!std::is_integral<El>::value, V>::type  // only real values are accepted
+    // operator()(const Container<Container<El, Allocator_inner>, Allocator_outer>& a,
+    //     const Container<Container<El, Allocator_inner>, Allocator_outer>& b) const;
+
```
Comment (jura-gresko) [inline modules/distance/k-random/VOI.hpp]:
```suggestion
```

File: modules/distance/k-random/VOI.hpp [hunk 4]
```diff
@@ -24,18 +23,25 @@ namespace metric {
...
  * @return value of entropy estimation of the data 
  */
-template <typename Container, typename Metric = metric::Euclidian<typename Container::value_type>, typename L = double>
-double entropy(
-    std::vector<Container> data, std::size_t k = 3, L logbase = 2, Metric metric = Metric());
+// template <typename Container, typename Metric = metric::Euclidian<typename Container::value_type>, typename L = double>
+// double entropy(
+//     std::vector<Container> data, std::size_t k = 3, L logbase = 2, Metric metric = Metric());
+
+// template <typename Container, typename Metric = metric::Euclidian<typename Container::value_type>, typename L = double>
```
Comment (jura-gresko) [inline modules/distance/k-random/VOI.hpp]:
```suggestion
```

File: modules/distance/k-random/VOI.hpp [hunk 5]
```diff
@@ -192,10 +214,14 @@ struct VOI_kl {
...
-    operator()(const Container<Container<El, Allocator_inner>, Allocator_outer>& a,
-        const Container<Container<El, Allocator_inner>, Allocator_outer>& b) const;
+    template <typename Container, typename T = type_traits::underlaying_type_t<Container>>
+    typename std::enable_if_t<!std::is_integral_v<T>, V>  // only real values are accepted
+    operator()(const Container& a, const Container& b) const;
+
+    // template <template <class, class> class Container, class Allocator_inner, class Allocator_outer, class El>
+    // typename std::enable_if<!std::is_integral<El>::value, V>::type  // only real values are accepted
+    // operator()(const Container<Container<El, Allocator_inner>, Allocator_outer>& a,
+    //     const Container<Container<El, Allocator_inner>, Allocator_outer>& b) const;
```
Comment (jura-gresko) [inline modules/distance/k-random/VOI.hpp]:
```suggestion
```

File: modules/distance/k-random/VOI.hpp [hunk 6]
```diff
@@ -228,10 +254,14 @@ struct VOI_normalized_kl : VOI_kl<V> {
...
```
Comment (jura-gresko) [inline modules/distance/k-random/VOI.hpp]:
```suggestion
```

File: tests/distance_tests/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,31 @@
+set(BOOST_TEST_LIB ${Boost_LIBRARIES})
```
Comment (jura-gresko) [inline tests/distance_tests/CMakeLists.txt]: there is something wrong with tabs & spaces for indent

**Comment by dv14** [comment]
applying @jura-gresko  suggestions

**Pull request merged**